Add OneCommon Dataset

takatoy opened the pull request:
**Patch description**
Adds the OneCommon dataset to the tasks.
AAAI 19 Paper: https://aaai.org/ojs/index.php/AAAI/article/view/4694
AAAI 20 Paper: https://arxiv.org/abs/1911.07588

**Testing steps**
```
python examples/display_data.py -t onecommon -dt train
python examples/display_data.py -t onecommon -dt valid
python examples/display_data.py -t onecommon -dt test
```

**Logs**
```
python examples/display_data.py -t onecommon -dt train
.
- - - - - - - - - - - - - - - - - - - - -
~~
[context]:
  0.46
  0.81
  -0.3333333333333333
  -0.5333333333333333
  -0.835
  ...and 23 more
[referents]:
  {'begin': 3, 'end': 7, 'target': [0, 0, 0, 1, 0, 0, 0]}
  {'begin': 9, 'end': 12, 'target': [0, 0, 0, 0, 1, 0, 0]}
  {'begin': 16, 'end': 16, 'target': [0, 0, 0, 1, 0, 0, 0]}
  {'begin': 19, 'end': 19, 'target': [0, 0, 0, 1, 0, 0, 0]}
[output]: None
[labels: do you have a black kinda small one with a very smaller one lighter directly to it 's right/below it ?]
~~
[context]:
  0.46
  0.81
  -0.3333333333333333
  -0.5333333333333333
  -0.835
  ...and 23 more
[referents]:
  {'begin': 1, 'end': 2, 'target': [1, 0, 0, 0, 0, 1, 0]}
  {'begin': 7, 'end': 10, 'target': [0, 1, 1, 0, 0, 0, 1]}
  {'begin': 21, 'end': 21, 'target': [1, 0, 0, 0, 0, 1, 0]}
  {'begin': 26, 'end': 29, 'target': [0, 0, 0, 0, 0, 1, 0]}
  {'begin': 31, 'end': 33, 'target': [1, 0, 0, 0, 0, 0, 0]}
[output]: None
[onecommon]: hmm i think the lower one is cut out for me . is there a very light grey dot up and to the left of the black one ?
[labels: ok that 's toward the bottom ? most of the others run in a line up and to the left above them ? if so pick the really light one above the black one]
~~
[context]:
  0.46
  0.81
  -0.3333333333333333
  -0.5333333333333333
  -0.835
  ...and 23 more
[referents]:
  {'begin': 6, 'end': 10, 'target': [0, 0, 0, 0, 0, 0, 1]}
[output]: 6
[onecommon]: no i 'm confused . do you have a large med grey dot with a smaller light grey dot down to its right ?
[labels: lol yes . let 's choose the large med grey one then , ty <selection>]
- - - - - - - - - - - - - - - - - - - - -
~~
[context]:
  -0.135
  -0.78
  0.3333333333333333
  -1.0
  0.64
  ...and 23 more
[referents]:
  {'begin': 2, 'end': 5, 'target': [1, 0, 0, 0, 0, 0, 0]}
  {'begin': 7, 'end': 7, 'target': [1, 0, 0, 0, 0, 0, 0]}
[output]: None
[labels: i have 1 pure black dot . it is largish in size]
~~
[context]:
  -0.135
  -0.78
  0.3333333333333333
  -1.0
  0.64
  ...and 23 more
[referents]:
  {'begin': 3, 'end': 4, 'target': [1, 0, 0, 0, 0, 0, 0]}
[output]: 0
[onecommon]: is it at the top ? there is a similar size grey one below it , with a line of 3 or 4 below that going from left to right upward ?
[labels: exactly . click the black <selection>]
- - - - - - - - - - - - - - - - - - - - -
~~
[ loaded 8304 episodes with a total of 46258 examples ]
```

```
python examples/display_data.py -t onecommon -dt valid
.
[ loaded 1038 episodes with a total of 5888 examples ]
```

```
python examples/display_data.py -t onecommon -dt test
.
[ loaded 1038 episodes with a total of 5710 examples ]
```

**Data tests (if applicable)**
```
Using /Users/takato/lab/project/ParlAI/.eggs/untokenize-0.1.1-py3.6.egg
running egg_info
writing parlai.egg-info/PKG-INFO
writing dependency_links to parlai.egg-info/dependency_links.txt
writing requirements to parlai.egg-info/requires.txt
writing top-level names to parlai.egg-info/top_level.txt
reading manifest file 'parlai.egg-info/SOURCES.txt'
reading manifest template 'MANIFEST.in'
writing manifest file 'parlai.egg-info/SOURCES.txt'
running build_ext
test_verify_data (test_new_tasks.TestNewTasks) ... ok

----------------------------------------------------------------------
Ran 1 test in 43.219s

OK
```

stephenroller commented:
Interesting, nice work! Thanks for making this!

Can you run `python parlai/scripts/verify_data.py -t onecommon` and report the results for me? The only thing that looks maybe off is it looks like some examples might have no `text` field (or rather, it's set to `None`). This can make it misbehave with most agents (https://github.com/facebookresearch/ParlAI/blob/6a7474fac9a4e5d6d3ac161c781ceade74acbe60/parlai/core/torch_agent.py#L1441-L1445). We solve this in other datasets by having a default `__SILENCE__` token that's used as the opener.

klshuster commented on parlai/tasks/onecommon/agents.py:
```diff
@@ -0,0 +1,268 @@
+        # Fill in teacher's message (THEM)
+        sentence = dialogue[entry_idx]
+        if sentence is None:
+            action['text'] = None
```
agree with @stephenroller it would be good to have a `__SILENCE__` token here

takatoy commented:
Thank you for reviewing!

Following your notes, I fixed the `None` part to `__SILENCE__`.
I also fixed a small bug which it was getting the wrong number of examples.

Below are the outputs of `verify_data.py` after all the fixes.

```sh
> python parlai/scripts/verify_data.py -t onecommon
[creating task(s): onecommon]
loading: /Users/takato/lab/project/ParlAI/data/onecommon/onecommon-master/aaai2020/experiments/data/onecommon/train_reference_0.txt
2s elapsed: {'exs': 7541, '%done': '32.60%', 'time_left': '4s', 'missing_text': 0, 'missing_labels': 0, 'missing_label_candidates': 7541, 'empty_string_label_candidates': 0, 'label_candidates_with_missing_label': 0, 'did_not_return_message': 0}
4s elapsed: {'exs': 15986, '%done': '69.12%', 'time_left': '1s', 'missing_text': 0, 'missing_labels': 0, 'missing_label_candidates': 15986, 'empty_string_label_candidates': 0, 'label_candidates_with_missing_label': 0, 'did_not_return_message': 0}
[ loaded 8304 episodes with a total of 23129 examples ]
5s elapsed: {'exs': 23129, '%done': '100.00%', 'time_left': '0s', 'missing_text': 0, 'missing_labels': 0, 'missing_label_candidates': 23129, 'empty_string_label_candidates': 0, 'label_candidates_with_missing_label': 0, 'did_not_return_message': 0}
```
```sh
> python parlai/scripts/verify_data.py -t onecommon -dt valid
[creating task(s): onecommon]
loading: /Users/takato/lab/project/ParlAI/data/onecommon/onecommon-master/aaai2020/experiments/data/onecommon/valid_reference_0.txt
[ loaded 1038 episodes with a total of 2944 examples ]
0s elapsed: {'exs': 2944, '%done': '100.00%', 'time_left': '0s', 'missing_text': 0, 'missing_labels': 0, 'missing_label_candidates': 2944, 'empty_string_label_candidates': 0, 'label_candidates_with_missing_label': 0, 'did_not_return_message': 0}
```
```sh
> pipenv run python parlai/scripts/verify_data.py -t onecommon -dt test
[creating task(s): onecommon]
loading: /Users/takato/lab/project/ParlAI/data/onecommon/onecommon-master/aaai2020/experiments/data/onecommon/test_reference_0.txt
[ loaded 1038 episodes with a total of 2855 examples ]
0s elapsed: {'exs': 2855, '%done': '100.00%', 'time_left': '0s', 'missing_text': 0, 'missing_labels': 0, 'missing_label_candidates': 2855, 'empty_string_label_candidates': 0, 'label_candidates_with_missing_label': 0, 'did_not_return_message': 0}
```

stephenroller commented:
The unit test failure is a false positive, you may ignore it.
